src/main.rs

```diff
@@ -256,231 +209,58 @@ async fn main() {
                 listener.local_addr().unwrap()
             );
             let _ = data_plane_port_tx.send(listener.local_addr().unwrap().port());
-            let active_client_count = Arc::new(AtomicUsize::new(0));
-            loop {
-                let listen_timeout =
-                    tokio::time::timeout(Duration::from_secs(10), listener.accept()).await;
-                if listen_timeout.is_err() {
-                    if active_client_count.load(Ordering::SeqCst) == 0 {
-                        info!("No active clients. Shutting down.");
-                        std::process::exit(0);
-                    }
-                    continue;
-                } else if let Ok(Ok((stream, _))) = listen_timeout {
-                    let conn = accept_connection(stream).await;
-                    {
-                        let mut r = IncrSvgDocServer::default();
-                        r.set_should_attach_debug_info(true);
-                        let client = Client {
-                            conn: Arc::new(Mutex::new(conn)),
-                            publisher: doc_publisher.clone(),
-                            doc_server: Arc::new(Mutex::new(r)),
+            while let Ok((stream, _)) = listener.accept().await {
+                let src_to_doc_rx = src_to_doc_jump.0.subscribe();
+                let world_tx = world_tx.clone();
+                let doc_watch_rx = doc_watch_rx.clone();
+                let conn = accept_connection(stream).await;
+                let actor::webview::Channels { svg, render_full } =
+                    actor::webview::WebviewActor::set_up_channels();
+                let render_tx = render_full.0.clone();
+                let webview_actor = actor::webview::WebviewActor::new(
+                    conn,
+                    svg.1,
+                    src_to_doc_rx,
+                    world_tx,
+                    render_full.0,
+                );
+                tokio::spawn(async move {
+                    webview_actor.run().await;
+                });
+                let render_actor =
+                    actor::render::RenderActor::new(render_full.1, doc_watch_rx, svg.0);
+                std::thread::spawn(move || {
+                    render_actor.run();
+                });
+                let mut renderer_rx = renderer_mailbox.0.subscribe();
+                tokio::spawn(async move {
```

Comment:
it is not possible for a non-async thread to listen two tokio channels at once. (listen on 1 channel is possible, by using blocking_recv.) Given that render_actor has two message source, I spawn a tokio task to as a "router" to combine and foward message for it.

Reply:
Intuitively, it is not the bottleneck in rust. Also, `comemo` requires to be binded to a specific thread, as discussed.